app/models/user.rb

```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  before_save {self.email = email.downcase}
+  validates :name, presence: true, length: {maximum: Settings.user.name.max_length}
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: {maximum: Settings.user.email.max_length},
+    format: {with: VALID_EMAIL_REGEX}, uniqueness: {case_sensitive: false}
+  has_secure_password
```

Comment:
has_secure_password là gì, và nó cho model User này những gì ?

Reply:
- has_secure_password được khai báo để bảo mật  password  người dùng. 
- Nó cho model User này.
     1. Nó có thể lưu password_digest an toàn vào trong cơ sở dữ liệu
     2. Nó yêu cầu xác thực password và password_confirmation phải trùng khớp giống nhau khi người dùng đăng ký
     3. Phương thức xác thực trả về cho người dùng khi mật khẩu là chính xác hoặc sai

Reply:
chưa hẳn chính xác những cũng tạm ổn rồi, bạn thử tìm hiểu xem nó cung cấp validate gì cho password không :v
Note: ko phải bảo mật mà là mã hóa nhé

Reply:
ok cảm ơn bạn. Theo mình biết thì has_secure_password nó cung cấp 1 validate cho password là : presence: true 
bắt trường hợp mật khẩu để trống

Reply:
ok, vậy thì cái presence true ở dòng dưới còn cần thiết không nhỉ, nếu theo như reply này của bạn thì đang bị lặp 2 lần validate presence này :grin:

Reply:
cái presence true ở dưới vẫn cần dùng đến. Vì nếu k có thuộc tính đó nó sẽ không kiểm tra password khi người dùng nhập vào là toàn khoảng trắng. Để tránh tình trạng lặp validate ta sẽ cho thêm thuộc tính allow_nil vào đó